From b0d2e97983b431f3c754d4af70b669a136897efd Mon Sep 17 00:00:00 2001
From: chdugw_auto <jkulkarni@maxlinear.com>
Date: Thu, 19 Aug 2021 09:32:31 +0300
Subject: [PATCH] Merge pull request #3481 in SW_UGW/linux from
 bugfix/PONRTSYS-9350-remove-bug_on-in-switch_api to master

* commit 'b3a1d2f36dc8f0da258f092167506f1824ef9d23':
  PONRTSYS-9350: GSW API - remove BUG_ON
---
 .../net/ethernet/lantiq/switch-api/gsw_flow_core.c | 22 ++++++++++++----------
 .../net/ethernet/lantiq/switch-api/gsw_flow_pce.c  |  3 ++-
 .../net/ethernet/lantiq/switch-api/gsw_tbl_rw.c    |  3 ++-
 3 files changed, 16 insertions(+), 12 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
index f7bf49060c48..d1b0b803f99a 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
@@ -2775,8 +2775,6 @@ u32 gsw_aging_get(void *cdev)
 	u32 exp = 0;
 	u64 tmp;
 
-	BUG_ON(!cdev);
-
 	if (!gswdev) {
 		pr_err("%s:%s:%d", __FILE__, __func__, __LINE__);
 		return GSW_statusErr;
@@ -2786,7 +2784,6 @@ u32 gsw_aging_get(void *cdev)
 		PCE_AGE_1_MANT_SIZE, &mant);
 	gsw_r32(cdev, PCE_AGE_0_EXP_OFFSET, PCE_AGE_0_EXP_SHIFT,
 		PCE_AGE_0_EXP_SIZE, &exp);
-	BUG_ON(!mant && !exp);
 
 	tmp = ((u64)mant + 1) << exp;
 	tmp *= gswdev->mactblsize;
@@ -2804,8 +2801,6 @@ int gsw_aging_set(void *cdev, u32 aging)
 	u32 tmp2;
 	u32 mant, exp;
 
-	BUG_ON(!cdev);
-
 	if (!gswdev || !aging)
 		return GSW_statusParam;
 
@@ -3567,7 +3562,10 @@ void *ethsw_api_core_init(ethsw_core_init_t *ethcinit)
 		gsw_w32(cdev, PCE_TFCR_NUM_NUM_OFFSET, PCE_TFCR_NUM_NUM_SHIFT,
 			PCE_TFCR_NUM_NUM_SIZE, (reg_val >> 2));
 
-		gsw_global_pcerule_bitmap_alloc(cdev);
+		if (gsw_global_pcerule_bitmap_alloc(cdev)) {
+			pr_err("failed in gsw_global_pcerule_bitmap_alloc\n");
+			return NULL;
+		}
 	}
 
 	if (!PrvData->aging_timeout)
@@ -27681,7 +27679,8 @@ static GSW_return_t GSW_GlobalRule_BitMapFree(void *cdev,
 {
 	ethsw_api_dev_t *gswdev = GSW_PDATA_GET(cdev);
 
-	BUG_ON(!cdev);
+	if (!cdev || !gswdev)
+		return GSW_statusErr;
 
 	bitmap_clear(gswdev->Global_Rule_BitMap.global_rule_idx_map,
 		   parm->base_index,  parm->num_of_rules);
@@ -27694,7 +27693,8 @@ static GSW_return_t GSW_GlobalRule_BitMapAlloc(void *cdev,
 {
 	ethsw_api_dev_t *gswdev = GSW_PDATA_GET(cdev);
 
-	BUG_ON(!cdev);
+	if (!cdev || !gswdev)
+		return GSW_statusErr;
 
 	parm->base_index =
 		bitmap_find_next_zero_area(gswdev->Global_Rule_BitMap.global_rule_idx_map,
@@ -27905,7 +27905,8 @@ GSW_return_t GSW_PceRuleEnable(void *cdev,
 	pctbl_prog_t pcetable;
 	int ret;
 
-	BUG_ON(!cdev || !gswdev);
+	if (!cdev || !gswdev)
+		return GSW_statusErr;
 
 	if (GSW_statusOk != gsw_pcerulemanage(cdev, parm))
 		return GSW_statusErr;
@@ -27941,7 +27942,8 @@ GSW_return_t GSW_PceRuleDisable(void *cdev,
 	pctbl_prog_t pcetable;
 	int ret;
 
-	BUG_ON(!cdev || !gswdev);
+	if (!cdev || !gswdev)
+		return GSW_statusErr;
 
 	if (GSW_statusOk != gsw_pcerulemanage(cdev, parm))
 		return GSW_statusErr;
diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_pce.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_pce.c
index 000cb90fa514..08f37a5ef955 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_pce.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_pce.c
@@ -1768,7 +1768,8 @@ int gsw_global_pcerule_bitmap_alloc(void *cdev)
 {
 	ethsw_api_dev_t *gswdev = GSW_PDATA_GET(cdev);
 
-	BUG_ON(!cdev);
+	if (!cdev || !gswdev)
+		return GSW_statusErr;
 
 	gswdev->Global_Rule_BitMap.global_rule_map_size =
 		 BITS_TO_LONGS(gswdev->num_of_global_rules) *
diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_tbl_rw.c b/drivers/net/ethernet/lantiq/switch-api/gsw_tbl_rw.c
index cb5828becfc0..ccac6cf070ee 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_tbl_rw.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_tbl_rw.c
@@ -836,7 +836,8 @@ GSW_return_t gsw_pce_rule_set(void *cdev, pctbl_prog_t *ptdata)
 	u32 ctrlval;
 	ethsw_api_dev_t *gswdev = GSW_PDATA_GET(cdev);
 
-	BUG_ON(!gswdev || !cdev);
+	if (!gswdev || !cdev)
+		return GSW_statusErr;
 
 	do {
 		gsw_r32_raw(cdev, PCE_TBL_CTRL_BAS_OFFSET, &ctrlval);
